""" rules for mapping reads to obtain RAD file """

rule map:
    input:
        idx = GENOME_DIR + "/splici_idx"
        tgmap = GENOME_DIR + "/splici_transcriptome/transcriptome_splici_fl86_t2g.tsv"
        read1 = DATA + "/preprocessed_fastqs/{sample}_R1.fastq.gz"
        read2 = DATA + "/preprocessed_fastqs/{sample}_R2.fastq.gz"
    output:
        mapped_data = OUT_DATA + "/mapped_reads"
    params:
        job_name = "map"
        memory = "select[mem > 45] rusage[mem = 45]", # LSF format; change as needed
        library_type = "ISR"
        end = 5
        barcodeLength = 16
        umiLength = 8
    log:
    threads:
        16
    shell:
        """
        salmon alevin \
            -l {params.library_type} \
            -1 {input.read1} \
            -2 {input.read2} \
            -i {input.idx} \
            -p {threads} \
            -o {output.mapped_data} \
            --tgmap {input.tgmap} \
            --end {params.end} \
            --barcodeLength {params.barcodeLength} \
            --umiLength {params.umiLength} \
            --keepCBFraction 1 \
            --sketch
        """